Terraform data source for vsphere_virtual_machine template results in panic

**Describe the bug**
I want to use vcsim to test out our Terraform VM deployment for VMware using templates uploaded in vcsim. 

**To Reproduce**
### Steps to Reproduce (on a Linux box)

1. Setup vcsim
```bash
$ curl -L -o - https://github.com/vmware/govmomi/releases/latest/download/vcsim_$(uname -s)_$(uname -m).tar.gz | sudo tar -C /usr/local/bin -xvzf - vcsim
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   151  100   151    0     0    164      0 --:--:-- --:--:-- --:--:--   164
100   631  100   631    0     0    501      0  0:00:01  0:00:01 --:--:--  616k
  0 10.2M    0  104k    0     0  10615      0  0:16:56  0:00:10  0:16:46 23202vcsim
100 10.2M  100 10.2M    0     0   964k      0  0:00:10  0:00:10 --:--:-- 2365k

$ vcsim version
Build Version: 0.26.1
Build Commit: ec5a4a7c
Build Date: 2021-08-17T02:48:00Z

$ vcsim
export GOVC_URL=https://user:pass@127.0.0.1:8989/sdk GOVC_SIM_PID=48828
```

2. At this point, open another terminal use this new terminal for rest of the instructions

3. Upload Template
```bash
$ govc version
govc 0.26.1

$ GOVC_INSECURE=1 GOVC_URL=https://user:pass@127.0.0.1:8989/sdk govc import.spec centos-7-7-template.ova | \
    jq '.NetworkMapping[0].Network = "VM Network" |
        .MarkAsTemplate = true |
        .Name = "centos-7-7-template"' > options.json
		
$ cat options.json
{
  "DiskProvisioning": "flat",
  "IPAllocationPolicy": "dhcpPolicy",
  "IPProtocol": "IPv4",
  "NetworkMapping": [
    {
      "Name": "bridged",
      "Network": "VM Network"
    }
  ],
  "MarkAsTemplate": true,
  "PowerOn": false,
  "InjectOvfEnv": false,
  "WaitForIP": false,
  "Name": "centos-7-7-template"
}

$ GOVC_INSECURE=1 GOVC_URL=https://user:pass@127.0.0.1:8989/sdk GOVC_RESOURCE_POOL=/DC0/host/DC0_C0 govc import.ova -options=options.json centos-7-7-template.ova
[31-08-21 04:25:52] Uploading template-disk1.vmdk... OK
[31-08-21 04:25:52] Marking VM as template...

$ GOVC_INSECURE=1 GOVC_URL=https://user:pass@127.0.0.1:8989/sdk govc ls /DC0/vm
/DC0/vm/DC0_H0_VM0
/DC0/vm/DC0_H0_VM1
/DC0/vm/DC0_C0_RP0_VM0
/DC0/vm/DC0_C0_RP0_VM1
/DC0/vm/centos-7-7-template
```
4. Create a terraform file `main.tf` with below contents:
```hcl
terraform {
  required_providers {
    vsphere = {
      source = "hashicorp/vsphere"
      version = "2.0.2"
    }
  }
}

provider "vsphere" {
  vsphere_server = "127.0.0.1:8989"
  user           = "user"
  password       = "pass"
}

data "vsphere_datacenter" "datacenter" {
  name = "DC0"
}

data "vsphere_compute_cluster" "cluster" {
  name          = "DC0_C0"
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_datastore" "datastore" {
  name          = "LocalDS_0"
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_network" "network" {
  name          = "VM Network"
  datacenter_id = data.vsphere_datacenter.datacenter.id
}

data "vsphere_virtual_machine" "template" {
  name          = "centos-7-7-template"
  datacenter_id = data.vsphere_datacenter.datacenter.id
}
```

5. Run Terraform
```bash
$ wget https://releases.hashicorp.com/terraform/1.0.5/terraform_1.0.5_linux_amd64.zip
--2021-08-31 05:49:59--  https://releases.hashicorp.com/terraform/1.0.5/terraform_1.0.5_linux_amd64.zip
Length: 32416786 (31M) [application/zip]
Saving to: ‘terraform_1.0.5_linux_amd64.zip’

terraform_1.0.5_linux_amd64.zip                     100%[======================================================>]  30.92M  11.3MB/s    in 2.7s

2021-08-31 05:50:13 (11.3 MB/s) - ‘terraform_1.0.5_linux_amd64.zip’ saved [32416786/32416786]

$ unzip terraform_1.0.5_linux_amd64.zip
Archive:  terraform_1.0.5_linux_amd64.zip
  inflating: terraform

$ terraform version
Terraform v1.0.5 (linux_amd64)

$ terraform init

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/vsphere versions matching "2.0.2"...
- Installing hashicorp/vsphere v2.0.2...
- Installed hashicorp/vsphere v2.0.2 (signed by HashiCorp)

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

$ VSPHERE_ALLOW_UNVERIFIED_SSL=true terraform plan
╷
│ Error: Plugin did not respond
│
│   with data.vsphere_virtual_machine.template,
│   on main.tf line 35, in data "vsphere_virtual_machine" "template":
│   35: data "vsphere_virtual_machine" "template" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadDataSource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-vsphere_v2.0.2_x5 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xf1c5c8]

goroutine 32 [running]:
github.com/hashicorp/terraform-provider-vsphere/vsphere/internal/virtualdevice.ReadNetworkInterfaces(0xc00015cb00, 0xe, 0x10, 0xc00057d350, 0x1, 0x1, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vsphere/internal/virtualdevice/virtual_machine_network_interface_subresource.go:579 +0x4e8
github.com/hashicorp/terraform-provider-vsphere/vsphere.dataSourceVSphereVirtualMachineRead(0xc0001fc500, 0x14351a0, 0xc000677180, 0x237d940, 0xc000580000)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vsphere/data_source_vsphere_virtual_machine.go:205 +0xa85
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0002fd5e0, 0x19e30a8, 0xc0004f8680, 0xc0001fc500, 0x14351a0, 0xc000677180, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:335 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc0002fd5e0, 0x19e30a8, 0xc0004f8680, 0xc000742a40, 0x14351a0, 0xc000677180, 0xc000677180, 0xc000742a40, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:558 +0xfd
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc00000c1b0, 0x19e30a8, 0xc0004f8680, 0xc000742220, 0xc0004f8680, 0x40b965, 0x16537a0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:1105 +0x4d6
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadDataSource(0xc00094e1a0, 0x19e3150, 0xc0004f8680, 0xc000716500, 0xc00094e1a0, 0xc00073f980, 0xc00057bba0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/server/server.go:247 +0xe5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler(0x16b5a40, 0xc00094e1a0, 0x19e3150, 0xc00073f980, 0xc000746420, 0x0, 0x19e3150, 0xc00073f980, 0xc000a20000, 0x4ba)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:416 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000c7500, 0x19ed138, 0xc000102c00, 0xc00015c500, 0xc0000ff050, 0x233e2d0, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/google.golang.org/grpc/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0000c7500, 0x19ed138, 0xc000102c00, 0xc00015c500, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/google.golang.org/grpc/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000038480, 0xc0000c7500, 0x19ed138, 0xc000102c00, 0xc00015c500)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/google.golang.org/grpc/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-vsphere/vendor/google.golang.org/grpc/server.go:857 +0x1fd

Error: The terraform-provider-vsphere_v2.0.2_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

**Expected behavior**
The Terraform plan should have not panicked.

**Affected version**
```bash
$ vcsim version
Build Version: 0.26.1
Build Commit: ec5a4a7c
Build Date: 2021-08-17T02:48:00Z

$ govc version
govc 0.26.1

$ terraform version
Terraform v1.0.5 (linux_amd64)
```

**Screenshots/Debug Output**
* [Terraform Debug](https://gist.github.com/kumarom/f9520a20af54f2b1c71bfcd684ea8dfa)
* [Terraform Panic](https://gist.github.com/kumarom/6c8943d3d0bef2dbe6493ef23ea4e23a)
* [Simulated VM Template Config](https://gist.github.com/kumarom/022be7e4f4e956f177e4ffa191a66fb5)

**Additional context**
* https://github.com/hashicorp/terraform-provider-vsphere/issues/1472
* #1779

--- a/simulator/virtual_machine.go
+++ b/simulator/virtual_machine.go
@@ -1114,6 +1114,17 @@ func (vm *VirtualMachine) configureDevice(ctx *Context, devices object.VirtualDe
 				Connected:      true,
 				DeviceConfigId: c.Key,
 			})
+
+			if c.ResourceAllocation == nil {
+				c.ResourceAllocation = &types.VirtualEthernetCardResourceAllocation{
+					Reservation: types.NewInt64(0),
+					Share: types.SharesInfo{
+						Shares: 50,
+						Level:  "normal",
+					},
+					Limit: types.NewInt64(-1),
+				}
+			}
 		}
 	case *types.VirtualDisk:
 		summary = fmt.Sprintf("%s KB", numberToString(x.CapacityInKB, ','))

